### PR TITLE
Cache purger now soft purges on content update events.

### PR DESCRIPTION
**Pull request by gklopper**
The website currently has a system that polls CAPI and soft purges (built before the event stream from CAPI went live).

I think this is a more effective place for it.

After this we are going to start ramping up cache times a bit.

(BTW - this is not tested yet, just some early morning coding form home)

### CODE REVIEW

File: src/main/scala/com/gu/fastly/Lambda.scala [hunk 1]
```diff
@@ -42,6 +44,35 @@ class Lambda {
    * @return whether a piece of content was purged or not
    */
   private def sendPurgeRequest(contentId: String): Boolean = {
+    val request = basicPurgeRequest(contentId).build()
+
```
Comment (johnduffell) [inline src/main/scala/com/gu/fastly/Lambda.scala]: tbh I'd have just added a parameter to sendPurgeRequest (..., soft: Boolean) and not doubled up on code

File: src/main/scala/com/gu/fastly/Lambda.scala [hunk 2]
```diff
@@ -23,8 +23,10 @@ class Lambda {
       (event.itemType, event.eventType) match {
         case (ItemType.Content, EventType.Delete) =>
           sendPurgeRequest(event.payloadId)
+        case (ItemType.Content, EventType.Update) =>
+          sendSoftPurgeRequest(event.payloadId)
         case other =>
-          // for now we only send purges for content takedowns, so ignore any other events
+          // for now we only send purges for content, so ignore any other events
```
Comment (johnduffell) [inline src/main/scala/com/gu/fastly/Lambda.scala]: out of interest, what else is there?

**Comment by johnduffell** [comment]
😴

File: src/main/scala/com/gu/fastly/Lambda.scala (hunk 2 above)
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/fastly/Lambda.scala]: @johnduffell At the moment we support event types of `Update`, `Delete` and `RetrievableUpdate`. You can see them here: https://github.com/guardian/content-api-models/blob/master/models/src/main/thrift/events/crier/event.thrift

@gklopper I think you will also want to include `RetrievableUpdate`. This is used when the content is too large and exceeds the kinesis limits. In the event of this, we change send this new event type with a payload containing the following: https://github.com/guardian/content-api-models/blob/master/models/src/main/thrift/events/crier/event.thrift#L24 (allowing the client to retrieve the item themselves). This typically occurs for live blogs, but is not exclusive of.

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/fastly/Lambda.scala]: @johnduffell @gklopper Out of interest, are there additional things that we could help you purge that we can't at current. I've always planned to enrich the types of events sent onto the firehose so I'm happy to take requests.

File: src/main/scala/com/gu/fastly/Lambda.scala [hunk 3]
```diff
@@ -23,8 +23,10 @@ class Lambda {
       (event.itemType, event.eventType) match {
         case (ItemType.Content, EventType.Delete) =>
           sendPurgeRequest(event.payloadId)
+        case (ItemType.Content, EventType.Update) =>
+          sendSoftPurgeRequest(event.payloadId)
```
Comment (mchv) [inline src/main/scala/com/gu/fastly/Lambda.scala]: We need to add following case as well:
 ```scala
(ItemType.Content, EventType.RetrievableUpdate) =>
   sendSoftPurgeRequest(event.payloadId)
```

File: src/main/scala/com/gu/fastly/Lambda.scala (hunk 2 above)
Comment (gklopper) [inline src/main/scala/com/gu/fastly/Lambda.scala]: @LATaylor-guardian what is the code for 'RetrievableUpdate'? is it 3? (for here https://github.com/guardian/fastly-cache-purger/blob/master/src/main/thrift/event.thrift#L7-L10)

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/fastly/Lambda.scala]: Ah, looks like the thrift file was copied into this project. You can delete it now and add a dependency on: https://github.com/guardian/content-api-models. Then you'll have the up to date file.

Comment (cb372) [inline src/main/scala/com/gu/fastly/Lambda.scala]: @gklopper Short answer: yes, it's 3. [Here](https://github.com/guardian/content-api-models/blob/master/models/src/main/thrift/events/crier/event.thrift) is the latest version of the file.

Long answer: When I wrote this Lambda we weren't publishing the Thrift file for Crier events to Maven, so I had to copy-paste the Thrift file into this project. But nowadays we are publishing it, as part of the `content-api-models` artifact. I'll make a separate PR to add a dependency on that and delete the copy-pasted file. For now, please just update the file manually.

Comment (gklopper) [inline src/main/scala/com/gu/fastly/Lambda.scala]: @LATaylor-guardian off the top of my head without thinking it through too much, things like an event on a tag when it is added to a piece of content. Things like RSS feeds are quite expensive and on a lot of our rss feeds (and likewise tag pages) we could have long cache times.

File: src/main/scala/com/gu/fastly/Lambda.scala [hunk 4]
```diff
@@ -42,6 +44,35 @@ class Lambda {
...
+
+  /**
+   * Send a soft purge request to Fastly API.
+   *
+   * See [[https://docs.fastly.com/guides/purging/soft-purges Fastly soft purging]].
+   *
+   * @return whether a piece of content was purged or not
+   */
+  private def sendSoftPurgeRequest(contentId: String): Boolean = {
+    val request = basicPurgeRequest(contentId)
```
Comment (adamnfish) [inline src/main/scala/com/gu/fastly/Lambda.scala]: I thought purges needed authenticated in Fastly?

Comment (johnduffell) [inline src/main/scala/com/gu/fastly/Lambda.scala]: it is, see https://github.com/guardian/fastly-cache-purger/pull/3/files/7b3e0bf8a016ae25325444b15a2c1decb6187bdf#diff-dec464a104a3e57383b9d6e6dd67cc52R82

Comment (adamnfish) [inline src/main/scala/com/gu/fastly/Lambda.scala]: #oops

**Comment by mchv** [comment]
testing this on `PROD` now with @gklopper

**Comment by gklopper** [comment]
Seems to work...

purger

File: src/main/scala/com/gu/fastly/Lambda.scala [hunk 5]
```diff
@@ -36,26 +41,34 @@ class Lambda {
   private val EmptyJsonBody: RequestBody =
     RequestBody.create(MediaType.parse("application/json; charset=utf-8"), "")
 
+  private sealed trait PurgeType
+  private object Soft extends PurgeType
+  private object Hard extends PurgeType
```
Comment (cb372) [inline src/main/scala/com/gu/fastly/Lambda.scala]: You could override `toString` in these two if you wanted slightly more eye-pleasing logging:

```scala
private object Soft extends PurgeType { override def toString = "soft" }
// ...
println(s"Sent $purgeType purge request")
```

**Comment by cb372** [comment]
Looks good to me, apart from a pedantic comment about logging. Feel free to ignore that.

**Comment by gklopper** [comment]
I would appreciate if someone could do the merge for me...

**Comment by mchv** [comment]
@gklopper I have now deployed `master`

**Pull request merged**